python/device/device.py

```diff
@@ -0,0 +1,314 @@
+        """
+        Thread to occasionally send telemetry containing information about how the test
+        is progressing
+        """
+        done = False
+
+        while not done:
+            # setting this at the begining and checking at the end guarantees one last update
+            # before the thread dies
+            if self.done.isSet():
+                done = True
+
+            props = {
+                "averageD2cRoundtripLatencyToGatewayInSeconds": self.metrics.d2c.latency.extract_average(),
+                "d2cInFlightCount": self.metrics.d2c.inflight.get_count(),
+                "d2cSuccessCount": self.metrics.d2c.succeeded.extract_count(),
+                "d2cFailureCount": self.metrics.d2c.failed.extract_count(),
+            }
+
+            msg = Message(json.dumps(props))
+            msg.content_type = "application/json"
+            msg.content_encoding = "utf-8"
+            logger.info("Sending thief telementry: {}".format(msg.data))
+            self.client.send_message(msg)
+
+            self.done.wait(self.config.system_telemetry_send_interval_in_seconds)
+
+    def update_thief_properties_thread(self):
+        """
+        Thread which occasionally sends reported properties with information about how the
+        test is progressing
+        """
+        done = False
+
+        while not done:
+            # setting this at the begining and checking at the end guarantees one last update
+            # before the thread dies
+            if self.done.isSet():
+                done = True
+
+            props = {
+                "d2cTotalSuccessCount": self.metrics.d2c.total_succeeded.get_count(),
+                "d2cTotalFailureCount": self.metrics.d2c.total_failed.get_count(),
+                "runState": self.metrics.run_state,
+            }
+            if self.metrics.run_end:
+                props["runEnd"] = self.metrics.run_end
+
+            logger.info("updating thief props: {}".format(props))
+            self.client.patch_twin_reported_properties(props)
+
+            self.done.wait(self.config.system_telemetry_send_interval_in_seconds)
+
+    def dispatch_c2d_thread(self):
+        """
+        Thread which continuously receives c2d messages throughout the test run.
+        This will be soon renamed to be the c2d_dispatcher_thread.
+        """
+        while not self.done.isSet():
+            msg = self.client.receive_message(timeout=1)
+
+            if msg:
+                obj = json.loads(msg.data.decode())
+                if obj.get("thiefHeartbeat"):
+                    logger.info("heartbeat received")
+                    self.last_heartbeat = time.time()
+                elif obj.get("thiefPingbackResponse"):
+                    list = obj.get("messageIds")
+                    if list:
+                        with self.d2c_set_lock:
+                            for message_id in list:
+                                self.d2c_confirmed.add(message_id)
```

Comment:
what is the difference between d2c_confirmed and d2c_unconfirmed, i.e. what do these lists mean?

Reply:
I'm going to rewrite this section of code.